### DISCUSSION TITLE
Big memory consume during initial serverpod's server startup

### DISCUSSION THREAD

**Discussion by sulthanalihsan** [serverpod/serverpod#1038] [General]
I have a problem when running the Serverpod server on my Ubuntu VPS. I encountered a problem where it fails to initialize properly. I noticed that a significant amount of memory is consumed during the initial startup, resulting in a "Killed" message. However, the memory usage is not as high after that.

This problem has become a dead-end as I'm unable to run my application at all. I've tried several approaches, such as adjusting the swap memory settings and removing some apps, but it doesn't work

How can I address this initialization issue? Is there a way to minimize the memory usage during the initial startup?

first time startup
![image](https://github.com/serverpod/serverpod/assets/30169687/6bb08d67-4379-4571-ba66-cc5b11124a92)

after a while
![image](https://github.com/serverpod/serverpod/assets/30169687/1a16cefa-b622-4468-a831-008575cdc03c)

killed message when running other app:
![image](https://github.com/serverpod/serverpod/assets/30169687/dd4e50c5-7a7c-4b45-bd99-9e07542f0a0b)

**Comment by vlidholt** [comment]
Hi! My guess is that the memory consumption is coming from the Dart compiler, are you running the server from source code? An option would be to either precompile the app with `dart compile` ahead of time to native code (for the fastest startup) or compile to kernel code (slightly slower).

You can read more here: https://dart.dev/tools/dart-compile

**Reply by sulthanalihsan** [reply]
thanks for your advice, I just realised and it really helpfull and significantly reduce the memory usage, btw I feel a little stupid with i was doing before, but thank you

one more question, I've read the documentation and from the many compile options available (exe, aot, jit, kernel) I feel a little confused which option I should choose for my case, can you give an advise which one is recommended for my case for running serverpod on a vps with ubuntu os? thanks you very much

**Reply by vlidholt** [reply]
I would use the `exe` option if possible, that should be the fastest to run. I think it requires that you build the executable on the same architecture as you are intending to run it on though. Second best option would be the kernel, if you for some reason cannot build the server on the same architecture.
